sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2102,6 +2102,55 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns a new Dataset by adding columns or replacing the existing columns that has
+   * the same names.
+   */
+  private[spark] def withColumns(colNames: Seq[String], cols: Seq[Column]): DataFrame = {
+    require(colNames.size == cols.size,
+      s"The size of column names: ${colNames.size} isn't equal to " +
+        s"the size of columns: ${cols.size}")
+    require(colNames.distinct.size == colNames.size,
```

Comment:
Also need to consider the case sensitivity issue.

Reply:
Good point.